### ISSUE TITLE
Rejecting CSS Modules in Workers

### ISSUE THREAD

**Issue by samsebree** [WICG/webcomponents#854]
Regarding [CSS Modules](https://github.com/w3c/webcomponents/blob/gh-pages/proposals/css-modules-v1-explainer.md):

We're currently working on upstreaming the [CSS Modules PR](https://chromium-review.googlesource.com/c/chromium/src/+/1799923).

It has been decided that CSS Modules should be rejected in workers. How exactly to perform that rejection is debatable.

One approach is to perform the rejection as a parse error. The proposal is to add a flag,"allow_document_only_module_type" , to [ModuleScriptFetcher::WasModuleLoadSuccessful](https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/loader/modulescript/module_script_fetcher.cc?q=ModuleScriptFetcher::WasModuleLoadSuccessful&sq=package:chromium&g=0&l=27). If WasModuleLoadSuccessful is called from a document, we will allow the CSS Modules to fetch, otherwise, throw a fetch error.

Another approach is to perform a check on the execution context during [ValueWrapperSyntheticModuleScript::CreateCSSWrapperSyntheticModuleScript](https://chromium-review.googlesource.com/c/chromium/src/+/1799923/15/third_party/blink/renderer/core/script/value_wrapper_synthetic_module_script.cc#25) and throw a parsing error if we're attempting to load the CSS modules from a context other than a document. 

The advantage of failing the fetch() based on MIME type might be that it would be clearer to the developer. The error message would be similar to the one you already get if you try to 'import' something that doesn't have a JS MIME type.

The parse error approach could be preferable architecturally speaking if the fetch() steps don't have to know about the module map's settings object.

Any opinions? @annevk @domenic @hiroshige-g @dandclark

**Comment by annevk** [comment]
I think the method by which to do this will greatly depend on the outcome of #839.

cc @littledan

**Comment by xtuc** [comment]
Indeed, this seems to fit well with https://github.com/littledan/proposal-module-attributes.

For instance:
```js
import sheet from "./example.css" as css;
```
would accept the CSS without the risk of `example.css` being a JavaScript file.

**Comment by littledan** [comment]
> It has been decided that CSS Modules should be rejected in workers.

Where could I learn more about this decision? I'm curious how it would affect other module types and other contexts that JS code runs in.

> I think the method by which to do this will greatly depend on the outcome of #839.

Agreed; That may give us extra spec mechanics to reject module graphs, which we could take advantage of/remain consistent with here.

**Comment by hiroshige-g** [comment]
> It has been decided that CSS Modules should be rejected in workers.

Er, I expected CSSStyleSheet can't be constructed on workers because   https://drafts.csswg.org/cssom/#cssstylesheet is exposed only to Window.
Are there attempts (e.g. in the context of Constructable Stylesheet) to enable constructing CSSStyleSheet on workers? (@rakina?). If so it would be possible to allow CSS Modules on workes as well.

**Comment by hiroshige-g** [comment]
Just to clarify (and align words to the spec e.g. https://html.spec.whatwg.org/multipage/webappapis.html#concept-script-parse-error):

1.

> The proposal is to add a flag,"allow_document_only_module_type" , to ModuleScriptFetcher::WasModuleLoadSuccessful. If WasModuleLoadSuccessful is called from a document, we will allow the CSS Modules to fetch, otherwise, throw a fetch error.

This is a fetch error approach (i.e. resolving with null module scripts, not a parse error that results in a non-null module script with a parse error).

2.

> Another approach is to perform a check on the execution context during ValueWrapperSyntheticModuleScript::CreateCSSWrapperSyntheticModuleScript and throw a parsing error if we're attempting to load the CSS modules from a context other than a document.

This is a parse error approach.
ValueWrapperSyntheticModuleScript::CreateCSSWrapperSyntheticModuleScript() is called when parsing and creating a module script, and exceptions there result in [parse error](https://html.spec.whatwg.org/multipage/webappapis.html#concept-script-parse-error).

---

> The parse error approach could be preferable architecturally speaking if the fetch() steps don't have to know about the module map's settings object.

+1. I prefer parse error approach (2.) for this reason. Fetch spec, fetch errors should preferrably depend only on fetch client's settings object.
Also, in terms of implementation, this would remove an extra flag passed across module loading layers.

> The advantage of failing the fetch() based on MIME type might be that it would be clearer to the developer. The error message would be similar to the one you already get if you try to 'import' something that doesn't have a JS MIME type.

The viewpoint of error messages sounds reasonable (I've missed it, thanks!).

In 2. (parse error approach), we could also argue that the error behavior is consistent.
For example, on workers,

- `import sheet from "./example.css" as css;`: fetch completes without fetch errors, and a parse error is thrown, no scripts are evaluated.
- `import sheet from "./example.js";` where its content is `const sheet = new CSSStyleSheet(...);` (a kind of polyfill of CSS modules): fetch completes without fetch errors, scripts are evaluated, and an evaluation error is thrown.

There are still a difference between parse and eval errors, but the fetch behavior is the same for both cases.

I feel these two cases above is conceptually similar (trying importing on CSS Modules or constructing CSSStyleSheet on wrong global scopes), more than to MIME type errors.

**Comment by hiroshige-g** [comment]
> > I think the method by which to do this will greatly depend on the outcome of #839.
> 
> Agreed; That may give us extra spec mechanics to reject module graphs, which we could take advantage of/remain consistent with here.

Maybe? (not so sure though). Do we want to wait for #839 to resolve before further discussion (and [implementation](https://chromium-review.googlesource.com/c/chromium/src/+/1799923)) around this issue?

**Comment by dandclark** [comment]
It's a good point that importing a CSS module in a Worker can be compared to trying to instantiate a constructable stylesheet, and I'm leaning towards the parse error approach based on this reasoning.

Although, this does seem subject to change depending on the outcome of #839 and I think any conclusion we might reach here would at least have to be revisited based on how that turns out.  For example say that we end up with a syntax that allows someone to write this sort of thing:
```import something from "./example" with type: typeThatDoesntExist```
Perhaps the mechanism that would fail the above should also be used to fail
```import sheet from "./example" with type: css``` when used in a Worker.

In any case I don't think it needs to block [this implementation](https://chromium-review.googlesource.com/c/chromium/src/+/1799923) as long as we're behind a flag and not really locked in to any decision.

**Comment by hiroshige-g** [comment]
> Although, this does seem subject to change depending on the outcome of #839 and I think any conclusion we might reach here would at least have to be revisited based on how that turns out. For example say that we end up with a syntax that allows someone to write this sort of thing:
> `import something from "./example" with type: typeThatDoesntExist`
> Perhaps the mechanism that would fail the above should also be used to fail
> `import sheet from "./example" with type: css` when used in a Worker.

Er, yes, #839 will affect how we reject non-CSS MIME types for `import sheet from "./example" with type: css`, which will affect the design/decision here.

> In any case I don't think it needs to block [this implementation](https://chromium-review.googlesource.com/c/chromium/src/+/1799923) as long as we're behind a flag and not really locked in to any decision.

Yeah, agree.

Let's land the implementation, resolve #839 and then resume the discussion here after that.
